src/GraphQL/Validation/ValidationContext.cs

```diff
@@ -204,31 +206,279 @@ public List<FragmentDefinition> GetRecursivelyReferencedFragments(Operation oper
+        private object GetVariableValue(IGraphType graphType, VariableDefinition variable, object input, IVariableVisitor visitor)
         {
-            Node = node;
-            Type = type;
+            return ParseValue(graphType, variable, variable.Name, input, visitor);
+
+            // Coerces a value depending on the graph type.
+            object ParseValue(IGraphType type, VariableDefinition variable, VariableName variableName, object value, IVariableVisitor visitor)
+            {
+                if (type is IInputObjectGraphType inputObjectGraphType)
+                {
+                    var parsedValue = ParseValueObject(inputObjectGraphType, variable, variableName, value, visitor);
+                    visitor?.VisitObject(this, variable, variableName, inputObjectGraphType, value, parsedValue);
+                    return parsedValue;
+                }
+                else if (type is NonNullGraphType nonNullGraphType)
+                {
+                    if (value == null)
+                        return ReportNullError(variable, variableName);
+
+                    return ParseValue(nonNullGraphType.ResolvedType, variable, variableName, value, visitor);
+                }
+                else if (type is ListGraphType listGraphType)
+                {
+                    var parsedValue = ParseValueList(listGraphType, variable, variableName, value, visitor);
+                    visitor?.VisitList(this, variable, variableName, listGraphType, value, parsedValue);
+                    return parsedValue;
+                }
+                else if (type is ScalarGraphType scalarGraphType)
+                {
+                    var parsedValue = ParseValueScalar(scalarGraphType, variable, variableName, value);
+                    visitor?.VisitScalar(this, variable, variableName, scalarGraphType, value, parsedValue);
+                    return parsedValue;
+                }
+                else
+                {
+                    throw new InvalidOperationException("The graph type is not an input graph type.");
+                }
+            }
+
+            // Coerces a scalar value
+            object ParseValueScalar(ScalarGraphType scalarGraphType, VariableDefinition variable, VariableName variableName, object value)
+            {
+                if (value == null)
+                    return null;
+
+                object ret;
+
+                try
+                {
+                    ret = scalarGraphType.ParseValue(value);
+                }
+                catch (Exception ex)
+                {
+                    throw new InvalidVariableError(this, variable, variableName, $"Unable to convert '{value}' to '{scalarGraphType.Name}'", ex);
+                }
+
+                if (ret == null)
+                    throw new InvalidVariableError(this, variable, variableName, $"Unable to convert '{value}' to '{scalarGraphType.Name}'");
```

Comment:
You are right. Fixed.

Reply:
> Also, what if it adds the error to the list and returns null, and then further assume this is an integer field of a object -- now ParseDictionary could throw an error because null is invalid for the field.

now this will not happen